sdks/java/core/src/main/java/org/apache/beam/sdk/util/RowJson.java

```diff
@@ -0,0 +1,240 @@
+          .put(BYTE, node -> (byte) node.asInt())
+          .put(INT16, node -> (short) node.asInt())
+          .put(INT32, node -> node.asInt())
+          .put(INT64, node -> node.asLong())
+          .put(FLOAT, node -> (float) node.asDouble())
+          .put(DOUBLE, node -> node.asDouble())
+          .put(BOOLEAN, node -> node.asBoolean())
+          .put(STRING, node -> node.asText())
+          .build();
+
+  private RowJson() { }
+
+  /**
+   * Creates a Jackson deserializer for a {@link Schema}.
+   */
+  public static Deserializer deserializerFor(Schema schema) {
+    return Deserializer.forSchema(schema);
+  }
+
+  /**
+   * Jackson deserializer for {@link Row Rows}.
+   */
+  public static class Deserializer extends StdDeserializer<Row> {
+
+    private Schema schema;
+
+    /**
+     * Creates a deserializer for a row {@link Schema}.
+     */
+    public static Deserializer forSchema(Schema schema) {
+      schema.getFields().forEach(Deserializer::verifyFieldTypeSupported);
+      return new Deserializer(schema);
+    }
+
+    private static void verifyFieldTypeSupported(Schema.Field field) {
+      Schema.FieldType fieldType = field.getType();
+      verifyFieldTypeSupported(fieldType);
+    }
+
+    private static void verifyFieldTypeSupported(Schema.FieldType fieldType) {
+      Schema.TypeName fieldTypeName = fieldType.getTypeName();
+
+      if (fieldTypeName.isCompositeType()) {
+        Schema rowFieldSchema = fieldType.getRowSchema();
+        rowFieldSchema.getFields().forEach(Deserializer::verifyFieldTypeSupported);
+        return;
+      }
+
+      if (fieldTypeName.isContainerType()) {
+        verifyFieldTypeSupported(fieldType.getComponentType());
+        return;
+      }
+
+      if (!VALUE_GETTERS.containsKey(fieldTypeName)) {
+        throw new IllegalArgumentException(
+            "ToRow transform does not support Row fields of type " + fieldTypeName.name());
+      }
+    }
+
+    private Deserializer(Schema schema) {
+      super(Row.class);
+      this.schema = schema;
+    }
+
+    @Override
+    public Row deserialize(JsonParser jsonParser, DeserializationContext deserializationContext)
+        throws IOException, JsonProcessingException {
+
+      JsonNode rootObject = jsonParser.readValueAsTree();
+
+      if (!rootObject.isObject()) {
+        throw new UnsupportedRowJsonException(
+            "Unable to convert '" + rootObject.asText() + "' to Beam Row, it is not a JSON object. "
+            + "Only JSON objects can be parsed to Beam Rows at the moment");
+      }
+
+      return nodeToRow(rootObject, schema);
+    }
+  }
+
+  private static Row nodeToRow(
+      JsonNode jsonNode,
+      Schema targetRowSchema) {
+    return
+        IntStream
+            .range(0, targetRowSchema.getFieldCount())
+            .mapToObj(i -> getFieldValue(jsonNode, targetRowSchema, i))
+            .collect(Row.toRow(targetRowSchema));
+  }
+
+  private static Object getFieldValue(
```

Comment:
nit: Following this is kind of confusing. It seems like this function should really part of your `mapToObj` call above.